### DISCUSSION TITLE
CastError when trying to remove Cloudinary upload files from a database

### DISCUSSION THREAD

**Discussion by K-Juno** [neuefische/questions#247] [Web] [answered]
Hey ya,

I need a tip please!

When trying to remove a file (previously uploaded on Cloudinary) from the database, I get the following error:

![Screenshot 2022-10-12 at 10 07 36](https://user-images.githubusercontent.com/109735507/195286664-ab0c27e1-daee-4a7b-bf07-5d795c0e59dc.png)

Since the asset_id is passed over in the upload process, it seems like that _**that**_ id (automatically?) is used for identifying which file is selected for removal.

```
const json = await response.json();
console.log(json);

const newAudio = {
id: json.asset_id,
src: json.secure_url,
};
```
However, the id is too long (according to the error message).
There should be a way to declare the ObjectId as the correct id??

Thanks for any kind of input!

Pull Request: https://github.com/K-Juno/capstone-project/pull/15

**Comment by pixelass** [comment, marked as answer]
In this file pages/api/audioFiles/[id].js

You delete byId but the ID is the id of the cloudinary asset.


```js
import dbConnect from '../../../dbConnect';
import Audio from '../../../models/Audio';

export default async function handler(req, res) {
  await dbConnect();

  if (req.method === 'DELETE') {
    const { id } = req.query;

    await Audio.findByIdAndDelete(id);

    res.status(200).json({ message: 'Audio deleted.' });
  }
}
```


You push new audio-files here:


```js
const newAudio = {
  id: json.asset_id,
  src: json.secure_url,
};

setAudioFiles([
  ...audioFiles,
  {
    newAudio,
  },
]);
```

then you save it to your DB

```js
await fetch('/api/audioFiles/create', {
  method: 'POST',
  headers: { 'Content-Type': 'application/json' },
  body: JSON.stringify(newAudio),
});
```

Instead you should first save to DB and then use the result.

```js
const newAudio = {
  id: json.asset_id,
  src: json.secure_url,
};

const fileInDB = await fetch('/api/audioFiles/create', {
  method: 'POST',
  headers: { 'Content-Type': 'application/json' },
  body: JSON.stringify(newAudio),
});

setAudioFiles([
  ...audioFiles,
  {
    newAudio: fileInDB,
  },
]);
```

SIDENOTE: it could be simplified to this: (not necessary and would probably cause additional changes)

```js
const fileInDB = await fetch('/api/audioFiles/create', {
  method: 'POST',
  headers: { 'Content-Type': 'application/json' },
  body: JSON.stringify(newAudio),
});

setAudioFiles([
  ...audioFiles,
  fileInDB // no extra surrounding object
]);
```

**Reply by K-Juno** [reply]
Hi Gregor, thanks so much for the instructions! I adjusted the codes as you said. It seems like I still need to define _id somewhere. I tried in the model / Schema, that was wrong.. XD

**Reply by pixelass** [reply]
The `_id` is generated from mongoDB. Yo CAN define it yourself

You can test this via `console.log(fileInDB)`

See: https://mongoosejs.com/docs/guide.html#_id

**Reply by K-Juno** [reply]
It works now since we went away from findById but use findOneAndDelete now.
Thanks again!! :D
